solutions/gold/ioi-06-pyramid.mdx

```diff
@@ -0,0 +1,125 @@
+---
+id: ioi-06-pyramid
+source: IOI 2006
+title: Pyramid
+author: Justin Ji
+---
+
+## Explanation
+
+Consider iterating over every possible upper left corner of our pyramid's base.
+We want to find the chamber inside our pyramid's base with the smallest sum of
+heights.
+
+This boils down to doing a 2D range minimum query on every possible upper left
+corner of our chamber. The sliding window approach is as follows: when iterating
+over each row, we maintain an array of monotonic deques for
+each column that allows us to find the best chamber for each column. Then,
+we keep another monotonic deque that keeps track of the best possible result
+from our sliding window of columns.
+
+Note that in order to achieve $\mathcal{O}(NM)$ time complexity, 2D prefix sums
+are required.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(NM)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main() {
+	int m, n, b, a, d, c;
+	cin >> m >> n >> b >> a >> d >> c;
+	vector<vector<int>> grid_sum(n + 1, vector<int>(m + 1));
+	for (int i = 1; i <= n; i++) {
+		for (int j = 1; j <= m; j++) {
+			cin >> grid_sum[i][j];
+			grid_sum[i][j] += grid_sum[i - 1][j] + grid_sum[i][j - 1] -
+			                  grid_sum[i - 1][j - 1];
+		}
+	}
+
+	/** @return sum of rectangle with corners at (r1, c1) and (r2, c2) */
+	const auto rect_sum = [&](int row_1, int col_1, int row_2,
+	                          int col_2) -> int {
+		return grid_sum[row_2][col_2] - grid_sum[row_2][col_1 - 1] -
+		       grid_sum[row_1 - 1][col_2] + grid_sum[row_1 - 1][col_1 - 1];
+	};
+	/** @return sum of pyramid rectangle with upper left corner at (r, c) */
+	const auto p_sum = [&](int row, int col) -> int {
+		return rect_sum(row, col, row + a - 1, col + b - 1);
+	};
+	/** @return sum of chamber rectangle with upper left corner at (r, c) */
+	const auto c_sum = [&](int row, int col) -> int {
```

Comment:
is the `-> int` needed

Reply:
i prefer having the trailing return, but i can remove it if u want

Reply:
no u right, type annotations never hurt